### ISSUE TITLE
encourage instead of discourage alias

### ISSUE THREAD

**Issue by rismoney** [PoshCode/PowerShellPracticeAndStyle#108]
https://github.com/PoshCode/PowerShellPracticeAndStyle/blob/master/Style-Guide/Naming-Conventions.md

Idiomatic powershell should be encouraged, and the only substantive reason for discouraging a particular alias is imcompatability across versions.  

Summary: use aliases everywhere and encourage people to use powershell in new and fun ways.

**Comment by lipkau** [comment]
My take on this is that **using** aliases is discuraged... not **defining** them

**Comment by rismoney** [comment]
I disagree.  Using aliases should be encouraged.

> The big reasons for discouraging aliases is readability and maintainability.  You can't guarantee anyone coming along later will know what a particular alias means and that will act as a barrier to others maintaining and using your code.

The versions of full powershell have a set of consistent built in aliases that are ....built in.  Example, % and ? should be used excessively and everywhere.  The vast community has this discouraged in scripts incorrectly.  The only reason to not use a specific subset of items, would be deprecation of aliases which is only a thing in the PS vs PSCore limited use scenario.  I'd challenge any other use case as fake news.

**Comment by lipkau** [comment]
I should have been more clear: _when writing a script, function or module_

**Comment by pauby** [comment]
> You can't guarantee anyone coming along later will know what a particular alias means

Complete agreement.

But you also can't guarantee that the alias you're using in your code is the same alias on my system. Aliases are designed for me to customise my environment to my liking for me. They are not for developers to use as a shortcut because they don't care about the end user.

**Comment by rismoney** [comment]
> You can't guarantee anyone coming along later will know what a particular alias means

Fallacy.  Get-help % is no different than get-help foreach-object.  If someone doesn't know what something means the help system is built for this.

> Also, use of aliases makes no sense if you're fully qualifying commands which is another thing we should be doing in scripts and functions.

The performance issue of alias resolution in all but edge cases I'd argue as overblown.  I will quantify it in a subsequent message for full disclosure but that's like saying - abandon PScmdlet for native dotnet functions since they'll  be faster than ps wrapped stuff.

Using Where-Object isn't safer than ? You can override Where-Object by just defining a func -- in fact, it's easier to override the cmdlet name than the alias. No -Force and -Option AllScope required.  So aliases are safer in this regard.  More FUD.

**Comment by OraDotNetDev** [comment]
As far as I was aware the use case for aliases is as convenience for [interactive use in the shell.](https://blogs.msdn.microsoft.com/powershell/2006/04/25/is-it-safe-to-use-aliases-in-scripts/) 

Built-in commands can be over-ridden. However this is much less common if not also bad practice and if anything could lead to the suggestion that you use Fully Qualified references to all your commands ie `Microsoft.PowerShell.Core\Get-Command` and there's definitely case that you should so that for your own modules.

Never-mind the confusion of PowerShell newbies searching Google for "Powershell %" and understanding what this construct means.

**Comment by pauby** [comment]
>  The vast community has this discouraged in scripts incorrectly.
> The only reason to not use a specific subset of items, would be deprecation of aliases which is only a thing in the PS vs PSCore limited use scenario.

That's a very valid reason for discouraging them. `PS vs PSCore` is not a `limited use scenario`

> I'd challenge any other use case as fake news.
> Fallacy
> More FUD.

You've had reasoned argument and then we get your responses above. You're not here for reasoned argument. You're here to impose your view and put down anybody else's who don't agree with you with comments like the above.

You're not listening and as such nobody is going to listen to you so your suggestions on aliases are not going to heard, they're not going to be taken seriously and they are not going to end in the 'best practice' guidelines being amended. As such we are all wasting our time in this thread.

There are ways and means of doing things and debating issues. This isn't it.

**Comment by rismoney** [comment]
I am listening and trying to convince the community to amend its stance.   I'm trying to have a discussion, but I wonder if there are impedements from attentional, availability cascade, and status quo biases preventing meaningful dialog versus keeping a completely open mind.

> That's a very valid reason for discouraging them. PS vs PSCore is not a limited use scenario

https://github.com/PowerShell/PowerShell/issues/5870

The fact here is that the community is not in agreement and therefore the ruling here should be to not preach practice and style that raises discourse.   The rules for interactive usage and scripts cannot be dissimilar any more than the rules for PS vs PSCore are.   You can't have the cake and eat it too.  

Here's an example - https://github.com/majkinetor/au/pull/19 . I'm not surprised @pauby, since you created https://github.com/majkinetor/au/pull/89

As for a meaningful run-time performance difference between alias and not:
I have found the difference between % and foreachobject across 2M iterations to still be well below subsecond - generally speaking - .5 a milli or less and sometimes and more often than not on smaller samples indeterminate.   If you are trying to optimize powershell for subsecond runtimes, I'd argue you are using the wrong tooling methods and would even question using cmdlets.  Plus you probably spend 3 seconds alone just typing ForEach-Object vs a single key!  Nonetheless, the point here being that performance is largely inconsequential and people who run into perf issues due to aliases is an extremely small edge case.  But I wouldn't swear them off any more than I would any other cmdlet that might have a faster dotnet counterpart.


```
$x=0
Measure-Command {(1..2000000) | foreach-object { $x++ } }
```

and

```
$x=0
Measure-Command {(1..2000000) | foreach-object { $x++ } }`
`

**Comment by rismoney** [comment]
I am not trying to put down anyone's point of view and apologize if I came across that way.  I just want to change this, and get rid of this rule altogether, and the negative trickle down effect it is having across PSSA and other quality projects.

**Comment by michaeltlombardi** [comment]
> Fallacy. Get-help % is no different than get-help foreach-object. If someone doesn't know what something means the help system is built for this.

The help system exists only in the prompt, not when reviewing scripts. In any case, leaving my code review to look up commands that are only unclear because they are aliases is not good UX.  To be very clear, no one is arguing aliases should be avoided at the prompt, only in scripts and function definitions which are meant to be maintained and/or shared, which is what this practice guide points to.

> You can override Where-Object by just defining a func -- in fact, it's easier to override the cmdlet name than the alias.

Which is more of an argument in favor of fully-qualified commands than using aliases.  Functions are generally only a problem when clobbered but people can and do treat aliases as personally mutable shortcuts.  Most people probably _don't_ override the most common aliases (`?`, `%`, etc) but in a script or function declaration clarity and lack of ambiguity are very desirable.

> As for a meaningful run-time performance difference between alias and not...

Your code example actually only executes `ForEach-Object` once in each test as it is called from the pipeline. I wish I had better on-hand data for you - aliases being slow was a subtopic at the PowerShell Summit, so someone _does_ have better data than me - but I don't. One thing to keep in mind is that it's not _just_ about performance - it's performance, safety, readability, and maintainability.

Aliases, given that context, are a losing proposition for scripts _but totally fine_ at the prompt.

**Comment by rismoney** [comment]
The code example was an oversight in my hastiness to serve an example.   Simply looping any alias vs a built in cmdlet should reveal a slight variation at very high usage..  I don't think anyone would argue this.

> The help system exists only in the prompt, not when reviewing scripts.
The help system that is available to you, depends entirely on the IDE of your choosing.  Every language has a learning curve.  

> people can and do treat aliases as personally mutable shortcuts

This is true, but the built-in's should be respected as Microsoft included them from original and have not removed them until core.   How people treat them, is on them.  If you choose to clobber an alias or a function, that is on you. 


> performance, safety, readability, and maintainability.
Lets review as I am not convinced that you gain any of these by avoiding aliases.  I think that is what is being debated.   There are many ways to do things in Powershell and I wouldn't make a commandment that says thou should only use dotnet functions over cmdlets because of perf.  Both have their place.

Safety?  I have already stated that aliases like functions can easily be overwritten with greater ease, thereby it's not safer.
```function Get-ChildItem {write-host "hello"}  # this is immediate clobber
set-alias gci -Value "write-host" #fails without -Force -Option AllScope
```

Readability?  This is hardly a true statement vs a matter of opinion.
* mount vs New-PSDrive? dir vs Get-ChildItem.  
* verbosity comes at a cost. both in file size, readability from long statements/scripts
* you lose the idiomatic coolness of the language.   I wish there were more things like ternaries, and shortcuts like `%` and `?` and even parameter shortcuts like `ea`.
* if the end user learned the ridiculousness of the operators (`-gt`, `-lt` and `-eq` being greater than, less than, and equal), then they obviously don't need the coddling you propose when using  select -eq select-object or where -eq where-object.
* when i see a cmdlet like Add-DnsServerResponseRateLimitingExceptionlist  I want a barf bag 

Maintainability - catchall... Aliases used in scripts are no harder to maintain than functions.  Microsoft released broken changes in Core, where tons of cmdlets don't work and the same goes for aliases.  A level of adaptation will be necessary to provide cross version compatibility which is independent of the context of alias usage.  I mean they broke AD in core.

**Comment by rkeithhill** [comment]
BTW early in the development of PSSA we postulated the need for different "PSSA profiles".  That is, the amount of "linting" needed for a personal module is lower than one you'd publish to co-workers which is less than one you'd publish to the PSGallery.  I could definitely understand how a set of rules tailored for publishing to the PSGallery would annoy folks that just want to publish a module for some co-workers to use.

This issue exists with C#'s FxCop / CodeAnalysis.  There are a ton of rules - many of which are opt-in.  Microsoft provides some high-level rule sets that you can pick from or you can create your own custom rule set:

![image](https://user-images.githubusercontent.com/5177512/39326558-a819d9ac-4952-11e8-8b5e-5cbcb07780f1.png)

**Comment by dlwyatt** [comment]
That’s how Pester does mocking of module-qualified calls, incidentally.

**Comment by Jaykul** [comment]
This is a deliberately broad statement, so stick with me:

## All types of aliases reduce readability

### Readability is certainly a matter of perception. 

What may seem readable to one person may not to another. Therefore, when we talk about readability we're striving to find the right balance between forcing people to learn a language's syntax and APIs (commands) and being able to intuit the right things about the code without full knowledge.

### But there are still absolutes

I believe the three following facts are constants when you consider all the types of aliases that are out there (or even just the first three types):

1. Aliases are more terse, and less self-explanatory than full command names.
2. Aliases increase concept count by introducing more names for the same things.
3. Nobody knows all the aliases.

Did you know `Remove-Item` has **six** built-in aliases?  Can you name them?

**Comment by Jaykul** [comment]
When we talk about readability we're not really talking about whether other PowerShell authors can _figure it out_ by looking it up. We're asking whether someone who knows Go or C# can _intuit_ what's going on...

I understand that for certain audiences, some aliases seem more readable than the full command names, but it actually depends very much on the context, and the audience.

`rmdir` is a **very** broadly recognized command. All your existing command-line users (regardless of their operating system of choice) will know it, and understand what it does. It's basically one of the best examples of readable aliases, so let's make it...

## An example to prove the rule: `rmdir`

Take the following code:

```posh
rmdir $Profile
```

How many of your readers can reconcile their "knowledge" of what they _think_ `rmdir` does, with the fact that it's actually an alias to a PowerShell command that works against folders, but **also** files?  How many of your readers will leap to the _incorrect_ conclusion that `$Profile` must point to my `$Home` directory, instead of a profile.ps1 script file?

On the other hand, do you think even a single one of those readers would be _**more** confused_ by this:

```posh
Remove-Item $Profile
```

The fact is that `Remove-Item` is plain English. _It's universally understood even by, say, Excel macro authors who've never used a shell -- even by non-technical people._ Some people still may not know whether `$Profile` is a file or a folder, but at least nobody is being _misled_ by their supposed knowledge of the alias.

So it's **not** really just a matter of opinion.

On top of that, no matter their backgrounds, because of _Type 3_ aliases, as I said before: nobody knows _all_ the aliases.

There's no value in making people look up `pumo` to discover that the authors of the module incorrectly used `pu` and that `mo` is a module.  If you write `Publish-Module`, anyone with a little familiarity with PowerShell will intuit what it does -- and if they can't, they can **[Google it](https://www.google.com/search?q=Publish-Module)** (which is something they cannot do with "pumo" -- go on, [try it](https://www.google.com/search?q=pumo), I'll wait).

**Comment by rismoney** [comment]
Agree 100% on context.

Rmdir to delete a profile is an intent obfuscation and i think everyone would agree a bad practice.  Contrast that with `del` where there is no ambiguity and readbility is higher from all angles is what I am calling into question.  That's not bad powershell where a rule should be written and a code violation raised.  

 For me the punctuation aliases in particular are the best and their counterparts shouldn't even be used.  ? is awesome and should flourish in code bases everywhere.   It's unfortunate the language took such a verbose take.  

Based on the community I am surprised people even use pipelining.  I mean why is `|` suddenly an easy lesson but `&` usage is out of bounds.

**Comment by Jaykul** [comment]
The point though, @rismoney, is that `Remove-Item .\notes.txt` is **more** readable: not just because linux uses `rm`, not `del` -- but because it's plain English. On top of that, it's more [**Googleable**](https://www.google.com/search?q=Remove-Item).

Even Linux command-line users don't use `del` (they use `rm`), so really, only someone who's used DOS would be guaranteed to recognize `del .\notes.txt` at all -- it's just _not_ more readable.

As you know, my personal opinion is that a PSScriptAnalyzer rule about aliases should just be _informational_ -- they are, as you say, not _bad_.

**Comment by Jaykul** [comment]
To flip the script for a moment...

## I don't think readability is the ultimate goal

Unless you're just blogging, and not writing code for a living.

The PowerShell Best Practices and Style Guide isn't about blogging. I mean, we want people to blog in such a way that they're supporting best practices, but the best practices aren't about blog readership, they're about the _maintainability of production code_.

Obviously maintaining code requires reading it, so readability is a part of this -- but in that sense we're concerned with coders who would be _able_ to maintain the code in question. If you're working on a module or other "library code" and production code -- you don't need to worry about _random blog readers_ who've never seen `%` and need to `Get-Help del`.

So although aliases reduce readability for people who are unfamiliar with PowerShell --and you definitely shouldn't use them on a blog post that's targeting beginner PowerShell users-- hopefully we can all agree that when working on production code we should be asking questions like:

1. How hard is it to maintain?
  a. How buggy is it?
  b. How easy it is to track down bugs?
  c. How hard is it to refactor?
2. How well does it work?
  a. How fast is it?
  b. How much RAM does it need?

At my company, when we've talked about readability I specifically say: "I don't care if a _newbie_ can't read it. I want to know, can the average new hire DevOps engineer read it?" Someone with Ops or programming background, and at least intermediate PowerShell knowledge (or extensive knowledge in some other scripting language, so we can expect them to come up to speed in PowerShell fairly quickly).

Make sense?

Well, it turns out that the folks who are coming in without PowerShell knowledge (that is, .NET developers, Python DevOps experts, etc) can read our code if it's written with full commands. They can figure out what it does. But when we use aliases _other than_ the **Type One** aliases I mentioned above (which are, as I mentioned, problematic for other reasons), they have problems. It takes longer for them to get it, longer to come up to speed.

So basically I just end up saying, look:

- the only cost of expanded aliases is a few extra characters in the file
- the expanded code is readable for a broader audience (although more verbose)
- we have tools that can "expand" aliases to their definitions
  - people who are reading my code can also use those tools themselves
  - (either way, authors don't actually need to _type_ more)

So I just don't have a problem with aliases. One way, or the other. As long as you document your module requirements, and don't use aliases you made up yourself 😉 

When it comes to _idiomatic_ PowerShell ...
When it comes to _performant_ PowerShell ... 

I'm less concerned with people using `pumo` and more concerned with getting them to use pipelines instead of loops, and writing functions with `process` blocks, that output **objects** (not hashtables), take their parameters `ValueFromPipelineByPropertyName` and have appropriate names or aliases on their parameters so they can match more pipeline input....

**Comment by darkoperator** [comment]
Correct. The main purpose of the guideline is not for when using PS in the Console but when writing code to server as the name implies a recommended guideline based on the experience of several in the community about to how better format and structure code to make it easier to read, maintainable and less likely to introduce errors.

**Comment by midacts** [comment]
I think of writing PowerShell in two categories.
- Interactive console
- Writing code

Aliases should be fine when using PowerShell interactively.
I do not think they should be using in code. Code should be more formal and easier to read when writing it for code.

Similar to trying to write something in the shortest one liner possible. This should be reserved for the console only.

**Comment by copdips** [comment]
> I think of writing PowerShell in two categories.
> 
> * Interactive console
> * Writing code
> 
> Aliases should be fine when using PowerShell interactively.
> I do not think they should be using in code. Code should be more formal and easier to read when writing it for code.
> 
> Similar to trying to write something in the shortest one liner possible. This should be reserved for the console only.

agree, alias only in the user profile for the direct console interaction.

**Comment by darkoperator** [comment]
Makes perfect sense in that context

**Comment by Guyver1wales** [comment]
I'll leave my two cents here for what its worth.
I have the same opinion about this as I have about documentation:
ITS NOT ABOUT YOU!!

running some one-liners = use whatever you like, knock yourself out.
putting together the skeleton of a new production script and just getting the guts together to confirm it'll do what you want = use whatever you like, knock yourself out.

Writing a Production script/module = ITS NOT ABOUT YOU!!

You could die the day after that script goes live, car crash driving to work, heart attack, whatever.
Some poor bast**d then has to pick up that script and figure out just what the hell you wrote.

So unless you're a cruel, selfish sadistic **** , just make the code as plain english as possible and as readable as possible.
The same applies to documentation. you dont document your systems to make YOUR life easier, you make it easier for the people around you and the poor sods who take your position after you when you decide to **** off somewhere else and leave no documentation.....

**Comment by rismoney** [comment]
cd > set-location

Sorry, not sorry.
